File: names.py
import sqlite3
import numpy as np
import scipy as sp
import matplotlib.pyplot as plt
import seaborn as sns
import pandas as pd

from fuzzywuzzy import fuzz
from fuzzywuzzy import process

sns.set_style("whitegrid")

con = sqlite3.connect("database.sqlite")
cursor = con.cursor()
cursor.execute("SELECT name FROM sqlite_master WHERE type='table';")

df = pd.read_sql_query("SELECT * FROM NationalNames", con)


def namePlt(name=None):
    '''
    Plot name usage over time by gender
    name = name string
    '''
    if name == None:
        print "gotta enter a name"
        return
    if len(name) > 1:
        name = name[0].upper() + name[1:len(name)].lower()
        name = "'" + name + "'"
        query = "SELECT * from NationalNames WHERE name = " + name
        nameDF = pd.read_sql_query(query, con)
        nameDF = nameDF.drop(['Id','Name'], axis=1)
    else:
        print "the name needs some more letters"
        return
    if len(nameDF.Gender.unique()) > 1: 
        fdf = nameDF.loc[nameDF['Gender'] == "F"]
        mdf = nameDF.loc[nameDF['Gender'] == "M"]
                
        fdf.columns = ['Year','Gender', 'F']
        mdf.columns = ['Year','Gender', 'M']

        nameDF = pd.merge(fdf, mdf, on='Year', how='outer', sort=True)
        
        plt.figure()
        plt.plot(nameDF['Year'], nameDF['F'], color='pink', label="F")
        plt.plot(nameDF['Year'], nameDF['M'], color='powderblue', label="M")
        plt.xlabel('Year')
        plt.ylabel('Count')
        plt.title(name)
        plt.legend()
        plt.show()
        
        return nameDF
    else:        
        if nameDF.Gender.unique() == 'F':
            lin_col = 'pink'
        else:
            lin_col = 'powderblue'

        plt.figure()
        plt.plot(nameDF['Year'], nameDF['Count'], color= lin_col)
        plt.xlabel('Year')
        plt.ylabel('Count')
        plt.title(name)
        plt.show()
        
    return nameDF
    

    



def nameVars(dat, name=None, numMatch=25):
    '''
    Get variants on a name by Levenshtein distance.
    dat = name data set
    name = name string
    numMatch = number of variants to return
    '''
    name = name[0].upper() + name[1:len(name)].lower()
    nsum = dat.groupby(['Name','Gender'])['Count'].sum()
    
    nsum = pd.DataFrame(nsum)
    nsum.reset_index(inplace=True)

    names = dat.Name.unique()

    ndf = process.extract(name, names, limit = numMatch)
    ndf = pd.DataFrame(data = ndf, columns = ['Name', 'sim'])

    ndf = pd.merge(ndf, nsum, on = 'Name', how= 'inner')

    return ndf 

File: names.R
library(RSQLite)
library(ggplot2)

db <- dbConnect(dbDriver("SQLite"), "database.sqlite")

namePlt <- function(babyName) {

   babyName = paste0(toupper(substr(babyName, 1, 1)), tolower(substr(babyName, 2, nchar(babyName))))
    bName = paste0("'", babyName, "'")

    ndat <- data.frame(dbGetQuery(db, paste0("SELECT *
                  FROM NationalNames
                  WHERE name = " , bName) 
                 ))
    
    if (length(unique(ndat$Gender)) > 1) {
        
        ff <- data.frame(Name = babyName, Gender="F", Year = 1880:2014)
        fdat <- subset(ndat, Gender == "F")
        fdat <- merge(fdat, ff, by=c("Name","Gender","Year"), all.y=T)
        
        mf <- ff
        mf$Gender <- "M"
        mdat <- subset(ndat, Gender == "M")
        mdat <- merge(mdat, mf, by=c("Name","Gender","Year"), all.y=T)
        
        ndat <- rbind(fdat, mdat)
        ndat[is.na(ndat)] <- 0
        
        p <- ggplot(ndat, aes(x=Year, y=Count, colour=Gender)) + geom_line(size=.75) + scale_colour_manual(values = c("pink", "powderblue"))+ ggtitle(babyName) + theme_bw()        
            
    } else {
            
        if (unique(ndat$Gender) == "F") {
            lcol = "pink"
        } else {
            lcol = "powderblue"
        }
        
        p <- ggplot(ndat, aes(x=Year, y=Count)) + geom_line(size=.75, colour=lcol) + ggtitle(babyName)+theme_bw()            
        
    }               
    
    print(p)
    return(ndat)
}


